pad_server/src/controller.c

```diff
@@ -100,8 +103,34 @@ static void controller_cleanup(void *arg) { controller_disconnect((controller_t
  * @param n The number of bytes to receive. Must be less than or equal to the length of `buf`.
  * @return The number of bytes read. 0 indicates no more bytes, -1 indicates an error and `errno` will be set.
  */
-static ssize_t controller_recv(controller_t *controller, void *buf, size_t n) {
-    return recv(controller->client, buf, n, 0);
+static ssize_t controller_receive(controller_t *controller, void *buf, size_t n) {
+    ssize_t bread = 0;
+    size_t total_read = 0;
+
+    while (total_read < n) {
+        bread = recv(controller->client, (char *)buf + total_read, n - total_read, 0);
+        if (bread <= 0) {
+            return bread;
+        }
+        total_read += bread;
+    }
+    return total_read;
+}
+
+/*
+ * Send buf to controller client
+ * @param controller The controller to send to.
+ * @param buf The buffer with the payload.
+ * @param n The size of buf
+ * @return The number of bytes written. -1 indicates an error.
+ */
+static ssize_t controller_send(controller_t *controller, void *buf, size_t n) {
+    return send(controller->client, buf, n, 0);
+}
+
+static ssize_t flush_recv(controller_t *controller) {
```

Comment:
To be so completely honest, I thought the flush option you presented was a standard library function that flushes the socket receive buffer itself, not a self-made implementation. My main concern with this implementation is that we could still get out of sync with the data stream. `RECV_BUFFER_FLUSH_SIZE` may not line up with the logical end of a packet.

Example:
We get a packet with malformed header. Payload could be any size according to us. For the sake of the example, the entire size is 16, but we don't know that.

The bad packet is followed by a good one of size 8, and another good one of size 10. 34 bytes total.

Let's say `RECV_BUFFER_FLUSH_SIZE` is 32 bytes. We flush out the 16 byte bad packet, the 8 byte good packet and then the first 2 bytes of the next good packet. We're left with it's remaining 8 bytes. On the next recv, it will appear to be another malformed header since we've accidentally flushed part of the header on the good packet.

This is just an example with example numbers, but it applies to all `RECV_BUFFER_FLUSH_SIZE` values. Packets can come in any length, so we can't make this number the greatest common multiple of packet lengths.

A working solution would read until the receive buffer is fully depleted, but that would require multiple small reads. In that implementation I worry that we continue to receive packets at the same rate or faster than we can flush with the small reads, getting us stuck flushing forever.

I think this might be outside the scope of this PR; you can save all this work on a separate branch to toy with maybe and brainstorm solutions. In the meantime I can merge this with the `sendmsg` changes and we can drop the flushing for now?

Sorry for the long read.